tests/test_validation_category_control_safety.py

```
@@ -0,0 +1,156 @@
+    assert _HostileScalarMeta.calls == 0
+    assert rust_calls == []
+
+
+def test_validate_judge_rejects_invalid_k_before_core_import(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """An invalid exact category count fails before compiled-core discovery."""
+    real_import = builtins.__import__
+    core_import_calls = 0
+
+    def guarded_import(name, globals=None, locals=None, fromlist=(), level=0):
```

Ruff (0.16.1)

[error] Function argument `globals` is shadowing a Python builtin

(A002)